Optimize `Connection::read_response`: Add caching

NobodyXu opened the pull request:
Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

NobodyXu committed:
Optimize `Connection::read_response`: Cache the whole packet if possible

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

NobodyXu committed:
Fix caching in `Connection::read_response`

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

jonhoo commented on src/connection.rs:
```diff
@@ -46,29 +46,44 @@ impl Connection {
+    pub(crate) async fn read_response(&mut self) -> Result<Response> {
+        // Clear content of buffer
+        //
+        // We assume that previous call to `Connection::read_response` does not
+        // read any bytes of the current packet to read in.
+        let buffer = self.transformer.get_buffer();
+        buffer.clear();
```
In that case wouldn't it be better to assert that the buffer is empty?

jonhoo commented on src/connection.rs:
```diff
@@ -46,29 +46,44 @@ impl Connection {
-        self.read_and_deserialize(len as usize).await
+        // Remove first 4b representing the len
+        let buffer = self.transformer.get_buffer();
+        buffer.drain(..4);
```
What's the type of `buffer` here? If it's not a `Bytes`, this could be quite expensive.

NobodyXu replied:
It is a Vec<u8> and that is indeed a bit expensive.

NobodyXu replied:
I guess it might be better to use `BytesMut` here to share the buffer between `Serializer` and `Deserializer` while making `drain` O(1).

NobodyXu replied:
I've fixed this by doing one `drain` before returning from `read_response`.

jonhoo commented on src/connection.rs:
```diff
@@ -46,29 +46,44 @@ impl Connection {
+        // Make sure we won't accidentally read in the next packet
+        // since `Connection::read_response` clear the buffer before
+        // reading anything.
+        debug_assert!(buffer.len() <= len);
```
I'm not sure I follow why we do it this way? Wouldn't it be better to not clear the buffer on each pass, and instead just consume `len` bytes from `buffer` and leave whatever trails it behind in the buffer for the next call?

Also, why is it okay for `buffer.len() < len`?

NobodyXu replied:
Because the same buffer is used for serializing.

> Also, why is it okay for buffer.len() < len?

The read_in_vec_rng reads at most 16 bytes, which is smaller than the maximum size of a packet.

NobodyXu replied:
Done!

NobodyXu commented:
After thinking through the problem, I think the best way to solve this is to update [`ssh_format`](https://github.com/openssh-rust/ssh-format) to support writing/reading directly to/from the stream and use [`BufStream`](https://docs.rs/tokio/latest/tokio/io/struct.BufStream.html) for buffering.

NobodyXu committed:
Rm `Session::try_wait`: None uses it and it just compliates `Connection`

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

NobodyXu committed:
Use separate buffers for read/write in `Connection`

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

NobodyXu commented:
@jonhoo I've rewritten the PR so that the read and write path use different buffers to avoid interfering with each other.

I also removed unused function `Session::try_wait` to simplify `Connection`.

jonhoo commented on src/connection.rs:
```diff
@@ -29,85 +29,53 @@ pub enum ForwardType {
-            }
-        }
+        // Remove the packet from buffer
+        buffer.drain(..(4 + packet_len));
```
This is still potentially going to be pretty expensive, since it requires shifting around all the remaining bytes of the `Vec`. You could try using a `VecDeque` instead, which doesn't have this problem, although then you have to deal with wrap-around, which may also be fairly painful.

NobodyXu replied:
That should be OK because most of the time, we receive one response packet after sending one request packet.

Only for a new session opened, can the mux server returns two packet in a row, but most of the time, it would take a long time before the second packet come through because the ssh mux server needs to communicate with the destination.
